featurebyte/query_graph/sql/interpreter/preview.py

```diff
@@ -138,7 +140,60 @@ def _apply_type_conversions(
 
         return sql_tree, type_conversions
 
-    def _construct_sample_sql(
+    def _get_query_graph_with_sample_on_primary_input_nodes(
+        self,
+        sample_on_primary_table: bool,
+        target_node_name: str,
+        seed: int,
+        sample_row_num: Optional[int] = None,
+        total_num_rows: Optional[int] = None,
+    ) -> tuple[QueryGraph, bool]:
+        """
+        Get query graph with sample row number set on primary input nodes
+
+        Parameters
+        ----------
+        sample_on_primary_table: bool
+            Whether to sample on primary table
+        target_node_name: str
+            Target node name
+        seed: int
+            Random seed
+        sample_row_num: Optional[int]
+            Number of rows to sample
+        total_num_rows: Optional[int]
+            Total number of rows before sampling
+
+        Returns
+        -------
+        Tuple[QueryGraph, bool]
+            Query graph with sample row number set on primary input nodes, whether input node is sampled
+        """
+        query_graph = QueryGraph(**self.query_graph.dict())
+        input_node_sampled = False
+        if not sample_on_primary_table:
+            return query_graph, input_node_sampled
+
+        has_join_node = False
+        for _ in query_graph.iterate_nodes(
+            target_node=query_graph.get_node_by_name(node_name=target_node_name),
+            node_type=NodeType.JOIN,
+        ):
+            has_join_node = True
+
+        if has_join_node and sample_row_num is not None and total_num_rows is not None:
+            sample_parameters = SampleParameters(
+                seed=seed, total_num_rows=total_num_rows, num_rows=sample_row_num
+            )
```

Comment:
Should we consider oversampling here and keep the final limit clause to handle the case where filters are applied to the view? Otherwise, the final result could return less than the desired number of rows.